scala/scala.bzl

```diff
@@ -21,360 +21,375 @@ _srcjar_filetype = FileType([".srcjar"])
-rm -rf {out}_args
-rm -rf {out}_tmp
-rm -rf {out}_tmp_expand_srcjars
-set -e
-mkdir -p {out}_args
-touch {out}_args/files_from_jar
-mkdir -p {out}_tmp""" + srcjar_cmd + """
-cat {scalac_args} {out}_args/files_from_jar > {out}_args/scala_args
-env JAVACMD={java} {scalac} {jvm_flags} @{out}_args/scala_args""" + javac_sources_cmd + """
-# add any resources
-{cp_resources}
-{java} -jar {jar} -m {manifest} {out} {out}_tmp
-rm -rf {out}_args
-rm -rf {out}_tmp
-rm -rf {out}_tmp_expand_srcjars
-""" + ijar_cmd
-  cmd = cmd.format(
-      cp_resources=cp_resources,
-      java=ctx.file._java.path,
-      jvm_flags=" ".join(["-J" + flag for flag in ctx.attr.jvm_flags]),
-      scalac=ctx.file._scalac.path,
-      scalac_args=scalac_args_file.path,
-      out=ctx.outputs.jar.path,
-      manifest=ctx.outputs.manifest.path,
-      jar=_get_jar_path(ctx.files._jar),
-      ijar=ctx.file._ijar.path,
+    jars = _jars
+    ijar_output_path = ""
+    ijar_cmd_path = ""
+    if buildijar:
+        ijar_output_path = ctx.outputs.ijar.path
+        ijar_cmd_path = ctx.file._ijar.path
+
+    java_srcs = _java_filetype.filter(ctx.files.srcs)
+    sources = _scala_filetype.filter(ctx.files.srcs) + java_srcs
+    srcjars = _srcjar_filetype.filter(ctx.files.srcs)
+    all_srcjars = set(srcjars + list(dep_srcjars))
+    # look for any plugins:
+    plugins = _collect_plugin_paths(ctx.attr.plugins)
+    plugin_arg = ",".join(list(plugins))
+
+    compiler_classpath = '{scalalib}:{scalacompiler}:{scalareflect}:{jars}'.format(  # noqa
+        scalalib=ctx.file._scalalib.path,
+        scalacompiler=ctx.file._scalacompiler.path,
+        scalareflect=ctx.file._scalareflect.path,
+        jars=":".join([j.path for j in jars]),
     )
-  outs = [ctx.outputs.jar]
-  if buildijar:
-    outs.extend([ctx.outputs.ijar])
-  ins = (list(jars) + 
-    list(dep_srcjars) + 
-    list(srcjars) + 
-    list(sources) + 
-    ctx.files.srcs + 
-    ctx.files.plugins + 
-    ctx.files.resources + 
-    ctx.files._jdk + 
-    ctx.files._jar + 
-    ctx.files._scalasdk + 
-    [ctx.outputs.manifest, 
-      ctx.file._ijar, 
-      ctx.file._scalac, 
-      ctx.file._java, 
-      scalac_args_file])
-  if compile_java_srcs:
-    ins.extend([javac_args_file])
-  ctx.action(
-      inputs=ins,
-      outputs=outs,
-      command=cmd,
-      mnemonic="Scalac",
-      progress_message="scala %s" % ctx.label,
-      arguments=[])
 
-def _compile_or_empty(ctx, jars, srcjars, buildijar):
-  # We assume that if a srcjar is present, it is not empty
-  if len(ctx.files.srcs) + len(srcjars) == 0:
-    _build_nosrc_jar(ctx, buildijar)
-    #  no need to build ijar when empty
-    return struct(ijar=ctx.outputs.jar, class_jar=ctx.outputs.jar)
-  else:
-    _compile(ctx, jars, srcjars, buildijar)
-    ijar = None
+    scalac_args = """
+JarOutput: {out}
+Manifest: {manifest}
+ScalacOpts: {scala_opts}
+Plugins: {plugin_arg}
+Classpath: {cp}
+Files: {files}
+EnableIjar: {enableijar}
+ijarOutput: {ijar_out}
```

Comment:
seems like maybe we should make this `IjarOutput` to be consistent, right? Same below.

Also, can we sort this list since order does not matter?